### PR TITLE
[NFC] Fix annoying build warnings and clang-tidy errors in ROCm 3.3

### PR DESCRIPTION
**Pull request by atamazov**
The list of fixes:
- clang-tidy: Globally disable "modernize-avoid-bind"
- clang-tidy: fix "constness of '...' prevents automatic move"
- clang-tidy: fix "error: method '...' can be made const"
- clang-tidy: fix "implicit conversion 'int' -> bool"
- clang-tidy: fix "method '...' can be made const".
- clang-tidy: fix "moving a temporary object prevents copy elision"
- clang-tidy: fix "narrowing conversion from '...' to ..."
- cppcheck: fix "Parameter '...' can be declared with const"

### CODE REVIEW

**Comment by atamazov** [comment]
I am urging to review this asap. I need this to use cppcheck and clang-tidy at full power on my system for comgr development.

File: src/include/miopen/handle.hpp [hunk 1]
```diff
@@ -227,23 +230,23 @@ struct Handle : miopenHandle
     }
 
 #if MIOPEN_USE_ROCBLAS
-    rocblas_handle_ptr& rhandle() { return rhandle_; }
+    const rocblas_handle_ptr& rhandle() const { return rhandle_; }
+    // rocblas_handle_ptr& rhandle() { return rhandle_; }
```
Comment (asroy) [inline src/include/miopen/handle.hpp:234]: why still keeping it?

**Comment by pfultz2** [comment]
> "prefer a lambda to std::bind"

You should just disable the tidy check globally in cmake file.

File: src/include/miopen/par_for.hpp [hunk 2]
```diff
@@ -91,9 +91,10 @@ void par_for_impl(std::size_t n, std::size_t threadsize, F f)
         const std::size_t grainsize = std::ceil(static_cast<double>(n) / threads.size());
 
         std::size_t work = 0;
-        std::generate(threads.begin(),
-                      threads.end(),
-                      std::bind(thread_factory{}, std::ref(work), n, grainsize, f));
+        std::generate(threads.begin(), threads.end(), [&]() {
+            return thread_factory()(work, n, grainsize, f);
+        });
+        // cppcheck-suppress unsignedLessThanZero
```
Comment (pfultz2) [inline src/include/miopen/par_for.hpp:97]: Why is this needed? It doesn't currently fail with cppcheck. Are you using a different version of cppcheck?

File: src/include/miopen/serializable.hpp [hunk 3]
```diff
@@ -82,24 +82,23 @@ struct Serializable
             ok = Parse<T>::apply(part, x);
         }
     };
+
+    // cppcheck-suppress constParameter
```
Comment (pfultz2) [inline src/include/miopen/serializable.hpp:86]: I think it would be better to keep the bind instead of using `cppcheck-suppress`.

File: src/include/miopen/par_for.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/miopen/par_for.hpp:97]: Ah wait, I think I might see the issue. Because you switch to using a lambda, cppcheck doesn't recognize how the `work` and `n` variable is changed.

File: src/include/miopen/handle.hpp (hunk 1 above)
Comment (atamazov) [reply src/include/miopen/handle.hpp:234]: typo

File: src/include/miopen/par_for.hpp (hunk 2 above)
Comment (atamazov) [reply src/include/miopen/par_for.hpp:97]: Exactly. I'll better disable tidy check globally. Actually I like this too. Exchanging one tidy warning to two cppcheck ones only to disable two instead of one looks weird))

**Comment by atamazov** [comment]
Thanks all reviewers, all comments should be resolved now.

**Comment by atamazov** [comment]
Description updated.

**Comment by pfultz2** [comment]
Thanks for fixing these!

**Comment by atamazov** [comment]
The change is truly NFC (unless compilers does not have errors).
